File: packages/sdk-android/src/main/kotlin/dev/useq/sdk/state/QState.kt
// Copyright (C) 2026 Ning Zou <q.cue.2026@gmail.com>
// SPDX-License-Identifier: AGPL-3.0-only

package dev.useq.sdk.state

import android.content.Context
import android.content.SharedPreferences
import java.util.UUID

/**
 * Lightweight persistence for Q SDK state using SharedPreferences.
 */
class QState(context: Context) {
    private val prefs: SharedPreferences =
        context.getSharedPreferences("dev.useq.sdk", Context.MODE_PRIVATE)

    companion object {
        private const val KEY_BUBBLE_X = "q_bubble_x"
        private const val KEY_BUBBLE_Y = "q_bubble_y"
        private const val KEY_SESSION_ID = "q_session_id"
    }

    // MARK: - Bubble position

    var bubbleX: Float
        get() = prefs.getFloat(KEY_BUBBLE_X, -1f)
        set(value) = prefs.edit().putFloat(KEY_BUBBLE_X, value).apply()

    var bubbleY: Float
        get() = prefs.getFloat(KEY_BUBBLE_Y, -1f)
        set(value) = prefs.edit().putFloat(KEY_BUBBLE_Y, value).apply()

    // MARK: - Session

    val sessionId: String
        get() {
            val existing = prefs.getString(KEY_SESSION_ID, null)
            if (existing != null) return existing
            val newId = UUID.randomUUID().toString()
            prefs.edit().putString(KEY_SESSION_ID, newId).apply()
            return newId
        }

    fun resetSession() {
        prefs.edit().remove(KEY_SESSION_ID).apply()
    }
}

File: packages/sdk-ios/Sources/QSDK/State/QState.swift
// Copyright (C) 2026 Ning Zou <q.cue.2026@gmail.com>
// SPDX-License-Identifier: AGPL-3.0-only

import Foundation

/// Lightweight persistence for Q SDK state using UserDefaults.
final class QState {
    private let defaults = UserDefaults(suiteName: "dev.useq.sdk") ?? .standard

    private enum Keys {
        static let bubbleX = "q_bubble_x"
        static let bubbleY = "q_bubble_y"
        static let sessionId = "q_session_id"
    }

    // MARK: - Bubble position

    var bubblePosition: CGPoint? {
        get {
            guard defaults.object(forKey: Keys.bubbleX) != nil else { return nil }
            return CGPoint(
                x: defaults.double(forKey: Keys.bubbleX),
                y: defaults.double(forKey: Keys.bubbleY)
            )
        }
        set {
            if let point = newValue {
                defaults.set(point.x, forKey: Keys.bubbleX)
                defaults.set(point.y, forKey: Keys.bubbleY)
            } else {
                defaults.removeObject(forKey: Keys.bubbleX)
                defaults.removeObject(forKey: Keys.bubbleY)
            }
        }
    }

    // MARK: - Session

    var sessionId: String {
        if let existing = defaults.string(forKey: Keys.sessionId) {
            return existing
        }
        let newId = UUID().uuidString
        defaults.set(newId, forKey: Keys.sessionId)
        return newId
    }

    func resetSession() {
        defaults.removeObject(forKey: Keys.sessionId)
    }
}
